__tests__/utilities.js

```diff
@@ -0,0 +1,396 @@
+// *** FOR JEST TESTING ***
+
+// Helper function for matching input objects with objects returned from DB queries
+const matchesDatabase = (inputObject, databaseObject) => {
```

Comment:
So, I know Allie touched on this in an earlier comment, but I'm going to make a suggestion:
So, it isn't *always* a problem to have a util function that you use a lot, but in this case it is a tricky enough function that it would probably be reasonable to write a test for the helper function. There are circumstances that call for that kind of thing, but it should raise a yellow flag that you may be doing more than is needed.

In this case, the heart of the problem, is that you are testing more than you need to, or specifically you are testing the same thing repeatedly.

It would be reasonable to have a single test per table, that makes sure that a row in that table stores and returns all of the values that we expect it to. but this should only be needed in a single test per table - and I would argue it would be better to write each of those tests out explicitly. Maybe with a couple smaller helpers just for date conversions.

This is because the thing that these tests should be checking for is that you have created the table correctly, and they protect against a future change that breaks the function you use for creating tables - HOWEVER, once you have confirmed that your function created the table as expected, you should stop testing that the table is correctly formatted. that is testing postgres instead of your own code.

All your other tests should just check the specific value or values that are relevant to them.

The problems with using the current approach are 1. this function is complicated enough to contain it's own bugs or get new ones if refactored 2. testing everything everywhere actually makes it harder to realize what is broken when it breaks. Because lets say the bug is that your table is missing a column - you would want just the "table has expected columns" test to fail instead of every single test.

So those would be my suggestions. That being said, I could imagine debating this with a coworker because there is an argument to be made for having this kind of DRY all-in-one test function that can be run at least once per table.

So if you want to keep this function as is, I would say write a test for it (with dummy input and outputs) and try to limit it's usage to one or two cases per table where it feels appropriate.

Reply:
Thank you for the feedback, Anthony. Before I read this, I had already begun the process of taking `matchesDatabase` out of my tests. So, I'll just keep doing that, but you make a lot of very good points that I will keep in mind in the future.